pkg/destroy/nutanix/nutanix.go

```diff
@@ -0,0 +1,211 @@
+}
+
+func cleanupVMs(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	matchedVirtualMachineList := make([]*nutanixclientv3.VMIntentResource, 0)
+	allVMs, err := o.v3Client.V3.ListAllVM(emptyFilter)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range allVMs.Entities {
+		if hasCategoryAssigned(v.Metadata, ec) {
+			matchedVirtualMachineList = append(matchedVirtualMachineList, v)
+		}
+	}
+
+	if len(matchedVirtualMachineList) == 0 {
+		o.logger.Infof("no VMs found that require deletion for cluster %s", o.clusterID)
+	} else {
+		logToBeDeletedVMs(matchedVirtualMachineList, o.logger)
+		err := deleteVMs(o.v3Client.V3, matchedVirtualMachineList, o.logger)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanupImages(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	allImages, err := o.v3Client.V3.ListAllImage(emptyFilter)
+	if err != nil {
+		return err
+	}
+
+	for _, i := range allImages.Entities {
+		imageName := *i.Spec.Name
+		imageUUID := *i.Metadata.UUID
+		if hasCategoryAssigned(i.Metadata, ec) {
+			o.logger.Infof("deleting image %s with UUID %s", imageName, imageUUID)
+			response, err := o.v3Client.V3.DeleteImage(imageUUID)
+			if err != nil {
+				o.logger.Errorf("failed to delete image %s: %q", imageUUID, err)
+				break
+			}
+
+			if err = nutanixtypes.WaitForTask(o.v3Client.V3, response.Status.ExecutionContext.TaskUUID.(string)); err != nil {
+				o.logger.Errorf("failed to confirm image deletion %s: %q", imageUUID, err)
+			}
+
+			break
+		}
+	}
+
+	return nil
+}
+
+func cleanupCategories(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	_, err := o.v3Client.V3.GetCategoryKey(ec.key)
+	if err != nil {
+		if strings.Contains(err.Error(), "does not exist") {
+			//Already deleted
+			return nil
+		}
+		return err
+	}
+
+	for _, val := range ec.values {
+		o.logger.Infof("deleting category value : %s", val)
+		err := o.v3Client.V3.DeleteCategoryValue(ec.key, val)
+		if err != nil {
+			o.logger.Errorf("failed to delete category value %s: %q", val, err)
+		}
+	}
+
+	o.logger.Infof("deleting category key : %s", ec.key)
+	return o.v3Client.V3.DeleteCategoryKey(ec.key)
+}
+
+func createExpectedCategory(infraID string) *expectedCategory {
+	return &expectedCategory{
+		key:    fmt.Sprintf(expectedCategoryKeyFormat, infraID),
+		values: []string{expectedCategoryValueOwned, expectedCategoryValueShared},
+	}
+}
+
+func logToBeDeletedVMs(tbd []*nutanixclientv3.VMIntentResource, l logrus.FieldLogger) {
+	l.Info("virtual machines scheduled to be deleted: ")
+	for _, v := range tbd {
+		l.Infof("- %s", *v.Spec.Name)
+	}
+}
+
+func deleteVMs(clientV3 nutanixclientv3.Service, vms []*nutanixclientv3.VMIntentResource, l logrus.FieldLogger) error {
+	taskUUIDs := make([]string, 0)
+	for _, vm := range vms {
+		l.Infof("deleting VM %s with ID %s", *vm.Spec.Name, *vm.Metadata.UUID)
+		response, err := clientV3.DeleteVM(*vm.Metadata.UUID)
```

Comment:
We need to save this error to return it after attempting to delete all of the VMs. As is, if there is a failure in a call to `DeleteVM`, the `deleteVMs` function still returns nil.